Implementation for Matching Engine Vectorstore

tomaspiaggio opened the pull request:
We just finished the implementation for the vector store using the GCP Matching Engine.

We'll be contributing the implementation.

Related to #2892

If you have any questions or suggestions please contact me (@tomaspiaggio) or @scafati98.

Commit:
Initial implementation for matching engine

Commit:
Fixing bugs

Commit:
Finished initial implementation. About to test

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
Typos

Commit:
fixes

scafati98 committed:
Merge branch 'master' of github.com:tomaspiaggio/langchain

Commit:
Documenting code. Added TODOs to scafati98

Commit:
Documenting code. Added TODOs to scafati98

Commit:
Added matchingengine.ipynb to documentation examples.

dev2049 commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,373 @@
+import time
+import uuid
+
+from google.cloud import aiplatform
```
all google imports should happen within constructor (or other class / instance methods). see `langchain/document_loaders/googledrive.py` for an example

dev2049 commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,373 @@
+        blob.upload_from_string(data)
+
+    def _get_gcs_client(self) -> storage.Client:
+        """Lazily creates a GCS client.
```
are there major benefits to lazy creation?

tomaspiaggio replied:
The benefit is that if the user does not use the `add_texts` method the client would never be created. I wouldn't call it a *major* benefit. @dev2049

vowelparrot replied:
I think it's better to require teh client in the class initialization, which also let's us collapse the credentials and project_id since they're contained within the client.

dev2049 commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,373 @@
+        index_id: str, # TODO document and tell the user that they need to run some lines to create the matching engine -> notebook de quickstart
+        endpoint_id: str, # TODO document and tell the user that they need to run some lines to create the matching engine -> notebook de quickstart
+        json_credentials_path: Union[str, None] = None,
+        embedder: Embeddings = TensorflowHubEmbeddings(model_url=HUB_MODEL)
```
convention in other vector stores is to name this variable `embedding`

dev2049 commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,373 @@
+    operation, updating the index takes close to one hour."""
+
+    def __init__(self,
+        project_id: str,
```
we're generally trying to move towards constructor only taking as input and storing repeatedly used objects, and everything that's only needed for initialization should be done outside constructor. and then if we want we can still have a convenience class method named something like `from_components` that puts together various id's and such into clients and indices and endpoints.

so in this case maybe that'd be somthing like
```python
def __init__(self, index, endpoint, client):
  ...
  
@classmethod
def from_components(project_id, region, gcs_bucket_uri, index_id, ...):
  index = _create_index_by_id(index_id)
  endpoint = ...
  client = ...
  return cls(index, endpoint, ...)
```

thoughts?

dev2049 reviewed (commented):
this is great! few comments

Commit:
Fixed the comments. Didn't fix linting yet. Stuck with Poetry.

tomaspiaggio commented:
I just pushed a new updates addressing the comments. However, we were trying to add `google-cloud-storage` and `google-cloud-aiplatform` to the `pyproject.toml` but we're having dependency conflicts with `black`. Do you have any suggetions here? @dev2049

dev2049 commented on pyproject.toml:
```diff
@@ -60,6 +60,7 @@ psycopg2-binary = {version = "^2.9.5", optional = true}
 pyowm = {version = "^3.3.0", optional = true}
 async-timeout = {version = "^4.0.0", python = "<3.11"}
 gptcache = {version = ">=0.1.7", optional = true}
+black = "^23.3.0"
```
think you may have accidentally added black here? shouldn't be a dependency

vowelparrot commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,465 @@
+              A valid gcs_bucket_uri or throws ValueError if full path is
+              provided.
+        """
+        gcs_bucket_uri = gcs_bucket_uri.replace("gs://", "")
```
If this is supposed to be a bucket name, it seems more straightforward to directly call the initialization argument "gcs_bucket_name". - does including the `gs://` prefix facilitate something I'm missing?

vowelparrot commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,465 @@
+    @classmethod
+    def _create_credentials_from_file(
+        cls,
+        json_credentials_path: Optional[str]
```
Looks like this is only called from the other classmethod in which the creds path is required. We may as well make it required here then

tomaspiaggio replied:
Actually, I ended up adding `Optional[str]` to `from_components` because the idea is that if the user doesn't pass any credentials, then the default will be used.

vowelparrot commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,465 @@
+        logger.debug(f"Embedding query {query}.")
+        embedding_query = self.embedding.embed_documents([query])
+
+        index_id = None
```
It would be cleaner to extract lines 321 to 329 as a separate method `_get_index_id()`

tomaspiaggio replied:
I like the suggestion

vowelparrot commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,465 @@
+        # means that the match method will always return an array with only
+        # one element.
+        for doc in response[0]:
+            print(doc)
```
```suggestion
```

vowelparrot commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,465 @@
+        gcs_bucket_uri: str = None,
+        index_id: str = None,
+        endpoint_id: str = None,
+        credentials_path: str = None,
```
If they're typed as `str` we shouldn't have a default of `None`. If we want these to be required keyword arguments, we could  do the patterh

```
def foo(*, region: str, index_id: str):
```

(the lone asterisk)

Commit:
Continued with comments

Commit:
Added dependencies to pyproject.toml

Commit:
Removed HUB_MODEL constant as it was the default anyway.

vowelparrot commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -0,0 +1,473 @@
+        Returns:
+            A configured MatchingEngine with the texts added to the index.
+        """
+        required_kwargs = ["project_id", "region", "gcs_bucket_uri",
```
If these are required, it's better to make them explicit. It seems better to do:
```python
    @classmethod
    def from_texts(
        cls: Type["MatchingEngine"], 
        texts: List[str],
        embedding: Embeddings = None,
        metadatas: Optional[List[dict]] = None,
        *,
        project_id: str,
        region: str,
        gcs_bucket_name: str,
...
```
etc.

Also - we should unify `gcs_bucket_uri` since it was schanged to `gcs_bucket_name` in the init (same as below)

If we're doing required kwargs here, should do required kwargs in from_components as well.


        **kwargs: Any,
    ) -> "MatchingEngine":

hwchase17 reviewed (commented):
my main comment, inline with some of the others - is it simpler to just do the client creation OUTSIDE of the class,  and then pass in an already initialized client? would cut back on a lot of the ags being passed around

tomaspiaggio commented:
@hwchase17 I thought that was addressed with the `from_components` function. Would you comment specifically what would you need? I'm also not sure what you mean by arguments being passed around as well. Would you please comment on that as well so I can fix it? Thank you!

Commit:
Changed gcs_bucket_uri to gcs_bucket_name in all places. Changed arguments in from_texts to make them required.

dev2049 commented:
> @hwchase17 I thought that was addressed with the `from_components` function. Would you comment specifically what would you need? I'm also not sure what you mean by arguments being passed around as well. Would you please comment on that as well so I can fix it? Thank you!

think he means to make `__init__` look something like what i mentioned here https://github.com/hwchase17/langchain/pull/3104/files#r1170476602

tomaspiaggio commented:
@dev2049 I already added the `from_components` function and I agree it is a better approach. The methods called in the constructor are validations for the `gcs_bucket_name` and that the client libraries are installed. I'm sorry if I'm not understanding what you mean.

Commit:
Moved all object creations out of the object and into from_components.

tomaspiaggio commented:
Great @dev2049 !! So do you need me to do anything else for the merge?

hwchase17 reviewed (approved):
looks great! thanks

meal commented:
@hwchase17 any chance to get this into release anytime soon?

eugenemiretsky commented:
@hwchase17 Same question here: Would be nice to see this released

eugenemiretsky commented:
One concern is that the docs are stored/retrieve from GCS which is slow (and somewhat defeats the purpose of using a Vector DB)

eugenemiretsky commented:
@tomaspiaggio should you create a  PR your branch to master?

tomaspiaggio commented:
Will this be merged to master? @hwchase17

HarrisonKhannah commented:
Keen to get this merged into master @hwchase17

ramssai commented:
Once we have Matching engine index is deployed, What is the best retriever on langchain to get the query results ? @tomaspiaggio

ktibbs9417 commented:
Have been using the Vector Search (Matching Engine) with langchain for a couple of days now and I've been hitting my head against a wall to solve a problem. 

I notice that when embeddings are sent to Vector Search they get stored and a file is also created and stored within a separate GCS bucket that is referenced when queried.

I am looking for a way to remove the embeddings from the Vector Search but it seems I can only do it with gcloud commands but I need to know the datapoint_ids. 

What would be the best way to store the datapoint_ids that are related to the documents that are being embedded?
